File: models/sketch_validation.py
"""Shared substantiveness heuristic for the learner's starting sketch.

This is the *only* place the substantiveness rule is defined for the backend.
Frontend (Plan B) ports this exact behavior to JS and verifies parity against
``tests/fixtures/sketch_validation_parity.json``.

A sketch is "substantive" when it carries enough learner-generated signal to
seed source-less provisional-map generation. The heuristic is deliberately
simple — token count + a small "don't know" pattern list — because:

  - It must run identically in two languages.
  - The cost of a false-negative (blocking a learner whose sketch is actually
    fine) is recoverable: the strategy-framed footer copy invites them to add
    more or attach source.
  - The cost of a false-positive (passing through a thin sketch that triggers
    hallucinated source-less generation) is foundational principle violation
    per spec §2 principle #7.

When in doubt, this returns False. Source attachment is always a valid
alternative for the learner.
"""
from __future__ import annotations

import re

# Minimum non-stopword token count to be considered substantive.
MIN_SUBSTANTIVE_TOKENS = 5

# Patterns the learner uses when they have nothing to say. Matched as
# normalized substring of the *whole* normalized sketch (stripped, lowercased,
# punctuation collapsed). Keep the list short and obvious; longer/cleverer
# detection is a different problem.
_DONT_KNOW_PATTERNS = (
    "idk",
    "i dont know",
    "i don't know",
    "no idea",
    "no clue",
    "dunno",
    "not sure",
)

# A small English stopword set — kept tiny on purpose so we don't ship an
# external dictionary and so JS parity stays trivial.
_STOPWORDS = frozenset(
    """
    a an the and or but if of for in on at to from by with as is are was were
    be been being do does did has have had this that these those it its
    """.split()
)

_PUNCT_RE = re.compile(r"[^\w\s]")
_WHITESPACE_RE = re.compile(r"\s+")
_REPEATED_CHAR_RE = re.compile(r"^(.)\1{4,}$")  # aaaaa, ?????, ......


def _normalize(text: str) -> str:
    """Lowercase, strip, collapse whitespace, drop punctuation."""
    text = text.strip().lower()
    text = _PUNCT_RE.sub(" ", text)
    text = _WHITESPACE_RE.sub(" ", text)
    return text.strip()


def _is_dont_know(normalized: str) -> bool:
    """The whole sketch is essentially a 'don't know' pattern."""
    if not normalized:
        return True
    if _REPEATED_CHAR_RE.match(normalized):
        return True
    for pattern in _DONT_KNOW_PATTERNS:
        if normalized == pattern:
            return True
        if normalized.startswith(pattern + " "):
            extra = normalized[len(pattern) + 1:].split()
            if len(extra) <= 3:
                return True
    return False


def _count_substantive_tokens(normalized: str) -> int:
    """Token count after dropping stopwords and very short tokens."""
    tokens = [t for t in normalized.split() if t and len(t) >= 2]
    return sum(1 for t in tokens if t not in _STOPWORDS)


def is_substantive_sketch(text: str) -> bool:
    """Return True if the sketch carries enough learner signal to seed
    source-less provisional-map generation.

    See the module docstring for the principle this enforces and why the
    heuristic is deliberately simple.
    """
    if text is None:
        return False
    normalized = _normalize(text)
    if not normalized:
        return False
    if _is_dont_know(normalized):
        return False
    if _count_substantive_tokens(normalized) < MIN_SUBSTANTIVE_TOKENS:
        return False
    return True

File: public/js/sketch-validation.js
// public/js/sketch-validation.js
//
// Substantiveness heuristic — JS port of models/sketch_validation.py.
// Verified against tests/fixtures/sketch_validation_parity.json by
// tests/test_frontend_sketch_validation.py. A divergence is a
// release-blocker per spec §5.3.
//
// Why /u everywhere: Python's \w is Unicode by default; JS's is ASCII-only.
// Without /u, the same input ("café résumé naïve …") would tokenize differently
// in the two languages and the parity fixture's unicode rows would silently
// fail. See models/sketch_validation.py "JS PORT NOTE" docstring.

export const MIN_SUBSTANTIVE_TOKENS = 8;

const DONT_KNOW_PATTERNS = [
  "idk",
  "i dont know",
  "i don't know",
  "no idea",
  "no clue",
  "dunno",
  "not sure",
];

const STOPWORDS = new Set(
  (
    "a an the and or but if of for in on at to from by with as is are was were " +
    "be been being do does did has have had this that these those it its"
  ).split(/\s+/)
);

const PUNCT_RE = /[^\p{L}\p{N}_\s]/gu;
const WHITESPACE_RE = /\s+/gu;
const REPEATED_CHAR_RE = /^(.)\1{4,}$/u;

function normalize(text) {
  let t = text.trim().toLowerCase();
  t = t.replace(PUNCT_RE, " ");
  t = t.replace(WHITESPACE_RE, " ");
  return t.trim();
}

function isDontKnow(normalized) {
  if (!normalized) return true;
  if (REPEATED_CHAR_RE.test(normalized)) return true;
  for (const pattern of DONT_KNOW_PATTERNS) {
    if (normalized === pattern) return true;
    if (normalized.startsWith(pattern + " ")) {
      const extra = normalized.slice(pattern.length + 1).split(/\s+/u).filter(Boolean);
      if (extra.length <= 3) return true;
    }
  }
  return false;
}

function countSubstantiveTokens(normalized) {
  const tokens = normalized.split(/\s+/u).filter((t) => t && t.length >= 2);
  let count = 0;
  for (const t of tokens) {
    if (!STOPWORDS.has(t)) count += 1;
  }
  return count;
}

export function isSubstantiveSketch(text) {
  if (text === null || text === undefined) return false;
  if (typeof text !== "string") return false;
  const normalized = normalize(text);
  if (!normalized) return false;
  if (isDontKnow(normalized)) return false;
  if (countSubstantiveTokens(normalized) < MIN_SUBSTANTIVE_TOKENS) return false;
  return true;
}
